echopop/graphics/plotting.py

```diff
@@ -0,0 +1,769 @@
+        ymin=length_labels.min() - length_offset,
+        ymax=length_labels.max() + length_offset,
+        colors="black",
+    )
+    # ---- Horizontal lines (length, y)
+    ax.hlines(
+        length_increments,
+        xmin=age_labels.min() - age_offset,
+        xmax=age_labels.max() + age_offset,
+        colors="black",
+    )
+
+
+def plot_age_length_distribution(
+    data_dict: Dict[str, Any],
+    **kwargs,
+):
+
+    # Get the dataset variable name
+    variable = kwargs.get("variable", None)
+
+    # Get sex
+    sex = kwargs.get("sex", None)
+
+    # Get the correct dataset for plotting
+    # ---- Abundance
+    if variable == "abundance":
+        dataset = data_dict["abundance"]["aged_abundance_df"].copy()
+        cmap = kwargs.get("cmap", "viridis")
+        colorbar_label = kwargs.get(
+            "colorbar_label", (f"Abundance ({sex} fish, # animals)").capitalize()
+        )
+    # ---- Biomass
+    elif variable == "biomass":
+        dataset = data_dict["biomass"]["aged_biomass_df"].copy()
+        cmap = kwargs.get("cmap", "plasma")
+        colorbar_label = kwargs.get("colorbar_label", (f"Biomass ({sex} fish, kg)").capitalize())
+
+    # Prepare the dataset
+    # ---- Stack sum
+    dataset_stk = dataset.stack(future_stack=True).sum(axis=1).reset_index(name=variable)
+    # ---- Convert the length and age columns from intervals into numerics
+    # -------- Length
+    dataset_stk["length"] = dataset_stk["length_bin"].apply(lambda x: x.mid).astype(float)
+    # -------- Age
+    dataset_stk["age"] = dataset_stk["age_bin"].apply(lambda x: x.mid).astype(float)
+    # ---- Subset the data based on the sex
+    dataset_sub = dataset_stk.loc[dataset_stk["sex"] == sex, :]
+
+    # Convert into a heatmap
+    heatmap_data = dataset_sub.pivot_table(
+        values=variable, columns="age", index="length", aggfunc="sum"
+    )
+
+    # Get limits
+    # ---- vmin
+    vmin = kwargs.get("vmin", 0.0)
+    # ---- vmax
+    vmax = kwargs.get("vmax", 10 ** np.round(np.log10(heatmap_data.stack().max())))
+
+    # Get axis limits
+    if "axis_limits" not in kwargs or kwargs.get("axis_limits") is None:
+        axis_limits = dict(
+            x=dict(xmin=dataset_sub["age"].min(), xmax=dataset_sub["age"].max()),
+            y=dict(
+                ymin=dataset_sub["length"].min() * 0.25, ymax=dataset_sub["length"].max() * 0.25
+            ),
+        )
+    else:
+        axis_limits = kwargs.get("axis_limits")
+
+    # Prepare default parameters
+    # ---- x
+    kwargs.update(dict(xlabel=kwargs.get("xlabel", "Age (years)")))
+    # ---- y
+    kwargs.update(dict(ylabel=kwargs.get("ylabel", "Fork length (cm)")))
+
+    # Initialize figure
+    # ---- Update the 'figsize' if it doesn't yet exist
+    if not kwargs.get("figsize", None):
+        kwargs.update(dict(figsize=kwargs.get("figsize", apply_aspect_ratio(5.5, axis_limits))))
+    # Initialize figure
+    fig, ax = plt.subplots(**prune_args(plt.subplots, **kwargs))
```

Comment:
What is the purpose of `prune_args`? Seems like it is to select only the parameters that subplot can accept?

Reply:
This is required because otherwise any and all `kwargs` supplied by the users error out for the various `matplotlib` functions. For instance, including `"figsize"`, `"vmin"`, and `"vmax"` in `kwargs` raises an Error when `kwargs` is passed to `matplotlib.pyplot.subplots` due to the inclusion of `"vmin"` and `"vmax"` despite `"figsize"` being a valid keyword argument. So `prune_args` subsets the parameters within `kwargs` that are a valid subset of those for the various functions where `prune_args` is used. I already tried to be as detailed as possible in the documentation for the `Survey.plot` method concerning these keyword args. I'll have to mull this over a bit more to think about alternative parameterizations.

Reply:
Yes, I know that this is to avoid users erroring out, but I don't think it's a good idea that the users actually put in erroneous arguments and do not know, and they would be left wondering why the command does not work.

I think the easiest way to solve this is to print a warning on anything that has been pruned because it is not allowed in matplotlib, and they will have no effects.

Reply:
Agreed. I'll add that to the checklist on things to incorporate!